### ISSUE TITLE
Parameter keys not work with minorDimension

### ISSUE THREAD

**Issue by walterxie** [CompEvol/beast2#969]
If `Parameter` represents a matrix, and `keys` can be the column/row names. But the current code `keys.length != dimension` makes this impossible.

**Comment by walterxie** [comment]
There is another problem in `Parameter` `String[] getKeys()`, where `Arrays.copyOf(...)` is used to ensure the `keys` not unmodifiable. But this will affect performance. I suggest to use `Collections.unmodifiableList` to `keys` and use `List<String> getKeys()` to return unmodifiableList `keys` directly.

**Comment by walterxie** [comment]
@rbouckaert as we discussed this morning, I fixed this bug and improve the performs of getKeys:

1. For 1D array, keys must have the same length as dimension. For 2D matrix, keys must either have the same length as dimension or the number of rows. 
2. Change keys to unmodifiable list `List<String> keys = Collections.unmodifiableList(...)`, which will throw UnsupportedOperationException if attempting to modify.
3. Related to 2, `String[] getKeys()` is Deprecated, and changed to `List<String> getKeysList()` where the list is the unmodifiable list.
